doc/markdownlinkconverter/markdownlinkconverter.c

```diff
@@ -0,0 +1,452 @@
+
+			//check target
+			bool targetOK = true;
+			//start
+			for (int i = 0;strcmp (ignoreTargetStart[i], "\0") != 0;++i)
+			{
+				if (strncmp (ignoreTargetStart[i], target,
+					strlen (ignoreTargetStart[i])) == 0)
+				{
+					targetOK = false;
+					break;
+				}
+			}
+			//end
+			for (int i = 0;strcmp (ignoreTargetEnd[i], "\0") != 0;++i)
+			{
+				if(len < strlen (ignoreTargetEnd[i]))
+					continue;
+
+				int j = len - strlen (ignoreTargetEnd[i]);
+				if (strncmp (ignoreTargetEnd[i], &target[j],
+					strlen (ignoreTargetEnd[i])) == 0)
+				{
+					targetOK = false;
+					break;
+				}
+			}
+			//print target
+			if (targetOK)
+			{
+				printTarget (output,target, filenameInElektra);
+			}
+			else fprintf (output, "%s", target);
+			fprintf (output, "%c", fgetc (input)); // print ")"
+		}
+		if (LINK_NAME (state, newstate))
+		{
+			++index;
+		}
+		if (LINK_TARGET (state, newstate)) ++len;
+		if (LINK_TRAP (state, newstate))
+		{
+			//trap, reset
+			if (fsetpos (input, &pos))
+				exitError (input, NULL, "fsetpos");
+
+			fprintf (output, "["); //first char got lost
+			state = LINK_START;
+			continue;
+		}
+		if(LINK_NOLINK (state, newstate))
+		{
+			// print all other content
+			fprintf (output, "%c", c);
+		}
+		state = newstate;
+	}
+}
+
+int main (int argc, char *argv[])
+{
+	if (argc < 2)
+	{
+		fprintf (stderr, "args error\n");
+		return EXIT_FAILURE;
+	}
+
+	char executablename[strlen (argv[0]) + 1];
+	char filename[strlen (argv[1]) + 1];
+
+	strcpy (executablename, argv[0]);
+	strcpy (filename, argv[1]);
+
+	char * filenameInElektra = getPathInElektraRoot (executablename, filename);
+
+	//1st pass
+	FILE * input = fopen (filename, "r");
+	if (!input)
+	{
+		fprintf (stderr ,"fopen Error\n");
+		return EXIT_FAILURE;
+	}
+	FILE * output = fopen (TEMP_FILENAME, "w+");
+	if (!output)
+	{
+		fclose (input);
+		fprintf (stderr ,"fopen Error\n");
+		return EXIT_FAILURE;
+	}
+
+	//save start of each file
+	fpos_t startInput;
+	fpos_t startTempFile;
+	if (fgetpos (input, &startInput))
+		exitError (input, output, "fgetpos");
+	if (fgetpos (output, &startTempFile))
+		exitError (input, output, "fgetpos");
+
+	if(!convertTitle (input, output, filenameInElektra))
+	{
+		/* No title found in file, therefore generate one and
+		 * print it out.
+		 */
+		//reset input file
+		if (fsetpos (input, &startInput))
+			exitError (input, output, "fsetpos");
+
+		fclose (output);
+		output = stdout;
+		//Generate Title
+		char * title = strrchr (filename, FOLDER_DELIMITER);
+		if (title == NULL)
+			exitError (input, NULL, "parsed file path invalid");
+
+		// print Title + Header
+		fprintf (output, "# %s # {#", &title[1]);
+		printConvertedPath (output, filenameInElektra);
+		fprintf (output, "}\n");
+		//2nd pass
+		convertLinks (input, output, filenameInElektra);
+		fclose (input);
+	} else
+	{
+		fclose (input);
+		//reset temp file
+		if (fsetpos (output, &startTempFile))
+			exitError (output, NULL, "fsetpos");
+
+		//2nd pass
```

Comment:
There can be markdown files with no headings and only at the end of a file you can be sure that there is no heading at all in a file. Adding the convention that every .md file must have at least one heading seemed harsh.